### ISSUE TITLE
Error message for missing return value with Result type is unclear

### ISSUE THREAD

**Issue by bguiz** [rust-lang/rust#24889]
The error message that I get looks like this:

```
src/main.rs:18:9: 18:11 error: unable to infer enough type information about `_`; type annotations required [E0282]
src/main.rs: 18         Ok(Response::with( (status::Ok, payload) ));
error: aborting due to previous error
```

The highlight is on the leading `Ok`, which is misleading. The actual fix was to either precede the statement with `return`, or to drop the trailing `;`. It would be great if the compiler were to suggest this as well, as the current error message alone is cryptic.

**Comment by bguiz** [comment]
CC @huonw

**Comment by huonw** [comment]
Code example:

```rust
fn foo() -> Result<u8, u64> {
    Ok(1);
}

fn main() {}
```

**Comment by bguiz** [comment]
Perhaps this should trigger `[E0269]` instead

**Comment by michaelsproul** [comment]
An ideal error message would be like the one generated here:

```rust
fn thing() -> u32 {
    7u32;
}

fn main() {}
```

```
test2.rs:1:1: 3:2 error: not all control paths return a value [E0269]
test2.rs:1 fn thing() -> u32 {
test2.rs:2     7u32;
test2.rs:3 }
test2.rs:2:9: 2:10 help: consider removing this semicolon:
test2.rs:2     7u32;
                   ^
error: aborting due to previous error
```

**Comment by steveklabnik** [comment]
Triage: here's today's error, no real change:

```text
error[E0282]: unable to infer enough type information about `_`
 --> <anon>:2:5
  |
2 |     Ok(1);
  |     ^^ cannot infer type for `_`
  |
  = note: type annotations or generic parameter binding required

```

**Comment by Mark-Simulacrum** [comment]
Going to go ahead and mark this as E-needstest since it feels easy to regress on and medium-level bad if we do.

In particular, I think adding to the existing UI test `src/test/ui/coercion-missing-tail-expected-type.rs` is the right thing to do. The example code given [here](https://github.com/rust-lang/rust/issues/24889#issuecomment-96980747) should be added to the existing UI test, the test run, and then the stderr file updated by cd-ing into `src/test/ui` and running the `update-all-references.sh` script. Given these instructions, marking as E-easy and E-mentor.

```
error[E0308]: mismatched types
 --> test.rs:1:29
  |
1 |   fn foo() -> Result<u8, u64> {
  |  _____________________________^
2 | |     Ok(1);
3 | | }
  | |_^ expected enum `std::result::Result`, found ()
  |
  = note: expected type `std::result::Result<u8, u64>`
             found type `()`
help: consider removing this semicolon:
 --> test.rs:2:10
  |
2 |     Ok(1);
  |          ^

error: aborting due to previous error(s)
```
